alf/experience_replayers/replay_buffer.py

```diff
@@ -0,0 +1,163 @@
+        with tf.device(self._device):
+            assert len(
+                env_ids.shape.as_list()) == 1, "env_ids should be an 1D tensor"
+            tf.Assert(batch_size == tf.shape(env_ids)[0], [
+                "batch and env_ids do not have same length", batch_size, "vs.",
+                tf.shape(env_ids)[0]
+            ])
+
+            # Make sure that there is no duplicate in `env_id`
+            _, _, env_id_count = tf.unique_with_counts(tf.sort(env_ids))
+            tf.Assert(
+                tf.reduce_max(env_id_count) == 1,
+                ["There are duplicated ids in env_ids", env_ids])
+
+            current_pos = tf.gather(self._current_pos, env_ids, axis=0)
+            indices = tf.concat(
+                [tf.expand_dims(env_ids, -1),
+                 tf.expand_dims(current_pos, -1)],
+                axis=-1)
+
+            tf.nest.map_structure(
+                lambda buf, bat: buf.scatter_nd_update(indices, bat),
+                self._buffer, batch)
+
+            self._current_pos.scatter_nd_update(
+                tf.expand_dims(env_ids, -1),
+                (current_pos + 1) % self._max_length)
+            current_size = tf.gather(self._current_size, env_ids, axis=0)
+            self._current_size.scatter_nd_update(
+                tf.expand_dims(env_ids, -1),
+                tf.minimum(current_size + 1, self._max_length))
+
+    def get_batch(self, batch_size, batch_length):
+        """Randomly get `batch_size` tranjectories from the buffer.
+
+        Note: The environment where the sampels are from are ordered in the
+            returned batch.
+
+        Args:
+            batch_size (int): get so many trajectories
+            batch_length (int): the length of each trajectory
+        Returns:
+            nested Tensors. The shapes are [batch_size, batch_length, ...]
+        """
+        with tf.device(self._device):
+            min_size = tf.reduce_min(self._current_size)
+            tf.Assert(min_size >= batch_length, [
+                "Not all environments has enough data. The smallest data "
+                "size is: ", min_size, "Try storing more data before "
+                "calling get_batch"
+            ])
+
+            batch_size_per_env = batch_size // self._num_envs
+            remaining = batch_size % self._num_envs
+            if remaining > 0:
+                # TODO: support this conidtion
+                tf.Assert(remaining == 0, [
+                    "batch_size=", batch_size, "cannot be divided by "
+                    "num_environments=", self._num_envs
+                ])
+
+            r = tf.random.uniform([self._num_envs, batch_size_per_env])
+            num_positions = self._current_size - batch_length + 1  # [num_envs]
+            num_positions = tf.expand_dims(num_positions, -1)  # [num_envs, 1]
+            pos = tf.cast(r * tf.cast(num_positions, tf.float32), tf.int32)
+            pos = tf.minimum(pos, num_positions - 1)  # [num_envs, K]
```

Comment:
Is this min necessary if tf.cast always rounds down?

Reply:
Because of floating multiplication at line 148, I am worried that it might be larger than num_positions - 1

Reply:
I think int(x*r) <= x-1 for any integer x and 0<r<1